CONTRIBUTING.md

```diff
@@ -29,3 +29,138 @@ Recommendations and requirements for how to best contribute to Parquet. We striv
+   the source of truth for the encoding).
+
+3. New compression mechanisms must have a pure Java implementation that can be
+   used as dependency in parquet-java.
+
+### Releases
+
+The Parquet PMC aims to do releases of the format package only as needed when
+new features are introduced. If multiple new features are being proposed
+simultaneously some features might be consolidated into the same release.
+Guidance is provided below on when implementations should enable features added
+to the specification.  Due to confusion in the past over parquet versioning it
+is not expected that there will be a 3.x release of the specification in the
+foreseeable future.
+
+### Compatibility and Feature Enablement
+
+For the purposes of this discussion we classify features into the following buckets:
+
+1. Backwards compatible. A file written under an older version of the format
+   should be readable under a newer version of the format.
+
+2. Forwards compatible. A file written under a newer version of the format with
+   the enabled feature can be read under an older version of the format, but
+   some information might be missing or performance might be suboptimal.
+
+3. Forward incompatible. A file written under a new version of the format with
+   the feature enabled cannot be read under an older version of the format (e.g.
+   Adding a new compression algorithm).
+
+New features are intended to be widely beneficial to users of Parquet, and
+therefore it is hoped third-party implementations will adopt them quickly after
+they are introduced. It is assumed that writing new parts of the format, and
+especially forward incompatible features, will be configured with feature flag
+defaulted to "off" and at some future point the features are turned on by default
+(reading of the new feature will typically be enabled without configuration or
+defaulted to on). Some amount of lead time is desirable to ensure a critical
+mass of Parquet implementations support a feature to avoid compatability issues
+across the ecosystem.  Therefore, the Parquet PMC gives the following
+recommendations for managing features:
+
+1. Backwards compatibility is the concern of implementations but given the
+   ubiquity of Parquet and the length of time it has been used, libraries should
+   support reading older version of the format to the greatest extent possible.
+
+2. Forward compatible features/changes may be used by default in implementations
+   once the parquet-format containing those changes has been formally released.
+   For features that may pose a significant performance regression to older
+   format readers, libaries should consider delaying default enablement until 1
+   year after the release of the parquet-java implementation that contains the
+   feature implementation.
+
+3. Forwards incompatible features/changes should not be turned on by default
+   until 2 years after the parquet-java implementation containing the feature is
```

Comment:
Still skeptical about a delay.

Reply:
Per above, I can understand the skepticism, I'd like to get more opinions here and we can update accordingly.

Reply:
We could add a shortcut like "or a PMC vote was in favour of it". If something gets really fast adoption, then we could enable it sooner.

How did you come up with "2 years"? Was this a guess or a deeper reasoning behind it?

Reply:
My suggestion is to avoid this sentence and just start with "It is recommended that changing the default value for a forward incompatible feature flag should be clearly advertised to consumers..."

Otherwise, this could be read as "any change made to parquet will not even plausibly be used for at least two years". I think it would be a shame if people used this 2 years as a reason to slow down their adoption of new parquet features

I think parquet would be better off if we let market demand drive the proliferation of features rather than trying to control a rollout. The thinking is that if new files begin appearing that are not readable by some system that is lagging, it would be beneficial to add pressure on that system upgrading their support

Reply:
> We could add a shortcut like "or a PMC vote was in favour of it". If something gets really fast adoption, then we could enable it sooner.

I think we should either do away with the guidance or release it.  PMC can always modify the language necessary via vote if necessary, but I'm not sure there is a good scenario where we want to actively break guidance?

> How did you come up with "2 years"? Was this a guess or a deeper reasoning behind it?

Mostly guesstimate and it might actually be too short.  This assumes a least some downstream dependencies only do major releases on a yearly basis as well.  So it leaves time for a downstream system to pickup a release in the following year and then ~1 year for rollout.   Taking Spark as an example, it appears to average about 3.5 years beetween major releases.  Minor releases are aimed at every six-months, but not everyone is going to use a release immediately.  If you look at [EMR](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-spark.html) ignoring the LTE release of 2.4.8, they are currently supporting 3.5.0 and 3.4.1. 3.4.1 was released  9 months ago and ~1 year ago respectively.   [Dataproc](https://cloud.google.com/dataproc/docs/concepts/versioning/dataproc-version-clusters), explicitly supports images for 24 months, but the actual Spark versions are older then AWS (i.e. 3.3)

If we do a parquet 2.0 release with a breaking change, we will probably learn how long it has taken people to actually adopt some of the newer breaking features.

> I think parquet would be better off if we let market demand drive the proliferation of features rather than trying to control a rollout. The thinking is that if new files begin appearing that are not readable by some system that is lagging, it would be beneficial to add pressure on that system upgrading their support

I think this ignores how long it takes changes to roll through the ecosystem, I detailed more in my reply on why 2 years above.  I'd welcome people to start experimenting features early but I think not recommending at least some delay will cause a lot of unnecessary pain.

Reply:
> I think this ignores how long it takes changes to roll through the ecosystem, I detailed more in my reply on why 2 years above

I am quite sure that stable mature enterprise platforms will take 2+ years to adopt any changes. However, this doesn't mean we should offer guidance that the entire ecosystem should follow that pace. 

Instead I think we (the parquet community) should offer help / visibility into how far the feature has percolated through the ecosystem (e.g. by a compatibility matrix) rather than some prescriptive proposal of a rollout schedule.

Reply:
> The thinking is that if new files begin appearing that are not readable by some system that is lagging, it would be beneficial to add pressure on that system upgrading their support

Having a not-before date published is a means of applying pressure without breaking anything.  I think it is much more reasonable messaging for downstream consumers to say, "you are seeing files that you can't read because you haven't updated your software for a long time".  It seems less reasonable to me to say to someone you are seeing "files you can't read because we a new feature was designed and implemented last month and we decided to make it default so  you should upgrade your software as well and/or contribute an implementation to the software you are using".  Also, maintainers have differing priorities, motivations and bandwidth.  "Pressuring" them because things are broken immediately rather than giving a reasonable time frame for implementation does not seem to me to be a considerate thing to do.

> I am quite sure that stable mature enterprise platforms will take 2+ years to adopt any changes. However, this doesn't mean we should offer guidance that the entire ecosystem should follow that pace.

Perhaps there is more word-smithing needed here.  This guidance is about when forward incompatible files should start being written by **default**. I expect that early adopters will opt into features sooner, especially if new features are appropriately publicized.  Early adopters are also usually more sophisticated and can better assess the risks for there use-case. There is already language here saying users can decide for themselves if they are sure they are using Parquet in a closed ecosystem (maybe it can be strengthened?).  I am more concerned about late adopters. I'd love to see new features "on" by default, but I just don't think it is serves the broader ecosystem well to do so.  As a  point of reference even two years might not be sufficient as seen by https://github.com/trinodb/trino/issues/7953 (part of this is likely partially caused over confusion of "V2").  

> Instead I think we (the parquet community) should offer help / visibility into how far the feature has percolated through the ecosystem (e.g. by a compatibility matrix) rather than some prescriptive proposal of a rollout schedule.

I think both are useful. A matrix doesn't help further adoption without a mechanism for guiding when a feature should be turned on by default.  I think it is especially difficult without more contextualized information about relative popularity of readers (e.g. if a parquet haskell implementation is lagging behind on a feature should it give pause for enablement). 

Ultimately, implementations not under the governance of the Parquet PMC are free to do what they want but I think providing official and consistent guidance here to give some lead time for breaking changes is in the best interest of everyone using Parquet.  

To annotate some of the options already suggested on the thread:
1.  Don't provide guidance.  I think this will result in either the current status quo which is to never turn anything on by default OR lead to many incompatibilities across the ecosystem.
2. PMC votes on when to recommend a feature be turned on (possibly informed by feature matrix?).  I don't think this provides consistency (or if there is a consistent  means of making decision we should simply document it here and avoid the overhead of votes).
3. Time based implementation/implementations.  Provides a consistent criteria that other implementations can understand and take action on.

Reply:
IMO, the 2-year guidance is better than other options. We don't want individual votes or discussions for action of each new feature. In addition, developers (of downstream projects or other Parquet implementation not governed under Parquet PMC) usually do not actively keep an eye on Parquet activities. The general 2-year guidance may give people enough time to take action to changes on the Parquet side.

Reply:
@julienledem ultimately whatever time period we pick is going to be arbitrary, but do you have a concrete suggestion.  

> We could have a setting to adjust this in environments where there is a single system reading/writing parquet: setForwardIncompatibleFeaturesOnByDefault(boolean)

This seems like a implementation specific thing but I'm not sure a global setting is the right thing to do here.  This is probably a discussion to have specifically around what we want to do in different parquet implementations.

Reply:
1. We should be explicit about after turning on by default libraries are encouraged to enable turning it off.
2. I think we should wait on specific policies that parquet-cpp and parquet-java will definitively turn these things.

Reply:
Updated to include recommendation for 1. 

For full deprecation I made a note we will update the document for further timeline.